crates/evm/src/precompile_cache.rs

```diff
@@ -0,0 +1,361 @@
+#[derive(reth_metrics::Metrics, Clone)]
+#[metrics(scope = "sync.caching")]
+pub(crate) struct CachedPrecompileMetrics {
+    /// Precompile cache hits
+    precompile_cache_hits: metrics::Gauge,
+
+    /// Precompile cache misses
+    precompile_cache_misses: metrics::Gauge,
+
+    /// Precompile cache size
+    ///
+    /// NOTE: this uses the moka caches' `entry_count`, NOT the `weighted_size` method to calculate
+    /// size.
+    precompile_cache_size: metrics::Gauge,
+}
+
+/// A custom precompile provider that wraps a precompile provider and potentially a cache for it.
+#[derive(Clone, Debug)]
+pub struct MaybeCachedPrecompileProvider<P> {
+    /// The precompile provider to wrap.
+    precompile_provider: P,
+    /// The cache to use.
+    #[cfg(feature = "std")]
+    cache: Option<alloc::sync::Arc<PrecompileCache>>,
+    /// The spec id to use.
+    spec: SpecId,
+    /// Cache metrics.
+    #[cfg(feature = "metrics")]
+    metrics: CachedPrecompileMetrics,
+}
+
+impl<P> MaybeCachedPrecompileProvider<P> {
+    /// Given a [`PrecompileProvider`]  and cache for a specific precompile provider,
+    /// create a cached wrapper that can be used inside Evm.
+    #[cfg(feature = "std")]
+    pub fn new_with_cache(
+        precompile_provider: P,
+        cache: alloc::sync::Arc<PrecompileCache>,
+    ) -> Self {
+        Self {
+            precompile_provider,
+            #[cfg(feature = "std")]
+            cache: Some(cache),
+            spec: Default::default(),
+            #[cfg(feature = "metrics")]
+            metrics: Default::default(),
+        }
+    }
+
+    /// Creates a new `MaybeCachedPrecompileProvider` with cache disabled.
+    pub fn new_without_cache(precompile_provider: P) -> Self {
+        Self {
+            precompile_provider,
+            #[cfg(feature = "std")]
+            cache: None,
+            spec: Default::default(),
+            #[cfg(feature = "metrics")]
+            metrics: Default::default(),
+        }
+    }
+}
+
+impl<CTX: ContextTr, P: PrecompileProvider<CTX, Output = InterpreterResult>> PrecompileProvider<CTX>
+    for MaybeCachedPrecompileProvider<P>
+{
+    type Output = P::Output;
+
+    fn set_spec(&mut self, spec: <CTX::Cfg as Cfg>::Spec) -> bool {
+        let old_spec = self.spec;
+        self.spec = spec.clone().into();
+
+        if self.precompile_provider.set_spec(spec) {
+            return true;
+        }
+
+        old_spec != self.spec
+    }
+
+    #[cfg(feature = "std")]
+    fn run(
+        &mut self,
+        context: &mut CTX,
+        address: &Address,
+        inputs: &InputsImpl,
+        is_static: bool,
+        gas_limit: u64,
+    ) -> Result<Option<Self::Output>, String> {
+        use revm::interpreter::{Gas, InstructionResult};
+
+        if let Some(cache) = &self.cache {
+            let key =
+                PrecompileKey { address: *address, spec: self.spec, input: inputs.input.clone() };
+
+            let cache_result = cache.cache.get(&key);
```

Comment:
I believe this is called per call, not limited to precompiles?

this overhead seems a bit much if we need to check  that the address is a precompile address first?

Reply:
yes good point, this is called per call. not sure if we can avoid it if the way to determine if an address is a precompile or not is trying to run it as a precompile, are there any other ways to check if the address is a precompile?

Reply:
we can avoid his because this is relatively large overhead per call

this is how the underlying precompiles does it:

```
        let Some(precompile) = self.precompiles.get(address) else {
            return Ok(None);
        };
```

which is just `.contains`

so what we should do here is short-circuit if the precompiles don't contain the address

Reply:
nice thx! done here https://github.com/paradigmxyz/reth/pull/15536/files/03e1f2a278ad38e09e827845786ce918b937211c..a95a58f94d02b2fc3c7f292a1dc3c405fe661966